tests/integration/test_charm_agent.py

```diff
@@ -0,0 +1,175 @@
+    await ops_test.model.wait_for_idle(status=ActiveStatus.name)
+
+    # Install specified Jenkins version
+    await app.units[0].ssh(
+        "sudo apt-get install -y --allow-downgrades jenkins={}".format(jenkins_version)
+    )
+    apt_cache_output = await app.units[0].ssh("apt-cache policy jenkins")
+    assert "Installed: {}".format(jenkins_version) in apt_cache_output
+
+    # Restart Jenkins
+    await app.units[0].ssh("sudo systemctl restart jenkins")
+
+
+@pytest_asyncio.fixture(
+    scope="module",
+    params=[
+        pytest.param(None, id="latest LTS jenkins version"),
+        # The following are the versions of Jenkins running in production in Canonical as of
+        # 2022-09-21
+        pytest.param("2.361.1", id="jenkins version 2.361"),
+        pytest.param("2.346.3", id="jenkins version 2.346"),
+        pytest.param("2.332.4", id="jenkins version 2.332"),
+        pytest.param("2.277.4", id="jenkins version 2.277"),
+        pytest.param("2.263.4", id="jenkins version 2.263"),
+        pytest.param("2.249.3", id="jenkins version 2.249"),
+        pytest.param("2.235.5", id="jenkins version 2.235"),
+        pytest.param("2.222.4", id="jenkins version 2.222"),
+        pytest.param("2.176.4", id="jenkins version 2.176"),
+        pytest.param("2.150.3", id="jenkins version 2.150"),
+    ],
+)
+async def app_jenkins_version(ops_test: OpsTest, app: Application, request: pytest.FixtureRequest):
+    """Install a range of jenkins versions to run the tests against."""
+    jenkins_version = request.param
+    if jenkins_version:
+        install_jenkins_version(ops_test=ops_test, app=app, jenkins_version=jenkins_version)
+
+    yield app
+
+
+@pytest_asyncio.fixture(scope="function")
+async def agent(ops_test: OpsTest):
+    """Deploy machine agent and destroy it after tests complete."""
+    agent: Application = await ops_test.model.deploy(
+        # Currently hardcoded due to the charm on charmhub not currently working, needs to be
+        # updated once a working version of the agent machine charm is deployed
+        "/home/jdkandersson/src/jenkins-agent-charm/"
+        "jenkins-slave_ubuntu-16.04-amd64_ubuntu-18.04-amd64_ubuntu-20.04-amd64.charm",
+        series="focal",
+    )
+    # Don't wait for active because the agent will start blocked
+    await ops_test.model.wait_for_idle()
+
+    yield agent
+
+    await agent.remove()
+    # Wait for machine to be de-provisioned, wait_for_idle doesn't work here because the machine
+    # goes missing and wait_for_idle never stops waiting
+
+    async def application_count():
+        """Count the number of applications in the model."""
+        return len((await ops_test.model.get_status()).applications)
+
+    for _ in range(120):
+        if await application_count() == 1:
+            break
+        await asyncio.sleep(1)
+    assert await application_count() == 1, "jenkins agent failed to be removed"
+
+
+@pytest_asyncio.fixture(scope="function")
```

Comment:
The scope of this fixture seems to be module. Same as `app_jenkins_version_2_150` and `agent_downloading_jnlp_file_related_to_jenkins`

Reply:
This fixture depends on the agent fixture which has `function` scope because the agent, if added, needs to be removed again after each test. That means that this dependent fixture also needs to have `function` scope. The agent fixture is needed because the agent charm needs to be deployed before it can be related to the server.

Reply:
Make sense